### PR TITLE
Fix label alignment bug in finetuning

### PR DESCRIPTION
**Pull request by matevosashot**
**Fix label alignment bug in finetuning**

Fixes label alignment issues in `sft_12hz.py` and `modeling_qwen3_tts.py` caused by incorrect interaction with `ForCausalLMLoss` from `transformers`.

**Context**

`ForCausalLMLoss` has two modes: passing `labels` automatically shifts them left by one (token *n* predicts *n+1*), while passing `shift_labels` uses them as-is.

**`sft_12hz.py`**

The old code manually shifted inputs and labels before passing to the talker (`inputs_embeds[:, :-1]`, `labels[:, 1:]`). This is unnecessary — `ForCausalLMLoss` already handles the left-shift internally. The fix passes full unshifted tensors and adjusts hidden state slicing accordingly. Also adds the missing `text_projection` call on text embeddings.

**`modeling_qwen3_tts.py`**

The subtalker outputs 15 codes per position. Passing them via `labels` causes `ForCausalLMLoss` to shift and drop one, leaving only 14 — misaligning with the 15 logit outputs. The fix passes subtalker labels via `shift_labels` instead, bypassing the automatic shift.

---
ForCausalLMLoss implementation:

```python
def ForCausalLMLoss(
    logits, labels, vocab_size,
    num_items_in_batch=None, ignore_index=-100,
    shift_labels=None, **kwargs,
) -> torch.Tensor:
    logits = logits.float()
    if shift_labels is None:
        # Shift so that tokens < n predict n
        labels = nn.functional.pad(labels, (0, 1), value=ignore_index)
        shift_labels = labels[..., 1:].contiguous()
    logits = logits.view(-1, vocab_size)
    shift_labels = shift_labels.view(-1)
    shift_labels = shift_labels.to(logits.device)
    loss = fixed_cross_entropy(logits, shift_labels, num_items_in_batch, ignore_index, **kwargs)
    return loss
```

### CODE REVIEW

**Comment by Zahgrom34** [comment]
Can confirm this fix with actual numbers. I was fine-tuning 12Hz-1.7B-Base on a low-resource language and got exactly the symptom from #39, loss looks healthy but output is pure alien noise. Spent days blaming my dataset lol.

So I ran a test that isolates the loss path without any training: generate audio with the stock model, then teacher-force its OWN codec-0 codes back through the training loss. A correct objective must score near zero on the model's own output. Random guessing over the 2048 codebook is ln(2048) = 7.62.

| loss path | CE on model's own codes |
|---|---|
| script as shipped (manual pre-shift + ForCausalLMLoss shifts again) | **9.07** |
| aligned, single shift (this PR) | **0.550** |

The shipped objective scores the model's own output WORSE than random. It literally trains "predict code t+2", and then inference feeds back under the t+1 contract, that's where the noise comes from.

After applying this fix (plus the sub-talker hidden state fix from #337, they need each other), the exact same data and config went from non-speech to natural prosody in one run. Also hit the speaker_encoder checkpoint issue from #222/#232 btw, all three are real.

Please merge this, half the open finetuning issues (#39, #178, #179) are probably this bug.

**Comment by iwannatrip02-sys** [comment]
Independent confirmation on **Qwen3-TTS-12Hz-0.6B-Base**. Setup:

- free Colab T4, fp32 with `sdpa`;
- 143 clips (≈ 12.6 min) of one Spanish speaker;
- lr 2e-5, effective batch 4.

**Teacher-forcing check.** I used the *base* model, before any training, on 8 real training clips (mean CE):

| | as shipped | aligned (this PR + #337) | uniform |
|---|---|---|---|
| talker (codec 0) | 8.82 | 1.61 | ln 2048 = 7.62 |
| sub-talker (codecs 1-15) | 10.66 | 5.38 | 7.62 |

**Fine-tuning on the same data:**

- **Shipped objective.** I only added the `text_projection` fix from #188 so that the 0.6B runs. The speech was
  intelligible but **not the target speaker**:
  - 3 of 4 takes came out with a female pitch: median F0 186–205 Hz, against the speaker's ~105 Hz;
  - speaker-encoder cosine against the speaker's own recordings was 0.92 (another recording of the same speaker
    scores 0.99);
  - the 8–12 kHz band was 15–25 dB below the training data.
- **Aligned objective.** I applied the talker shift fix and the sub-talker `shift_labels` fix from this PR, plus the
  hidden-state index fix from #337. **Every take has the target speaker's voice**:
  - cosine 0.979–0.995;
  - median F0 94–123 Hz, except runaway generations;
  - the high band is back in the data's range.

**Remaining issue: text fidelity.** On this small dataset it is still unstable. On 4 unseen sentences × 2 takes, the
checkpoints after epochs 2, 4 and 6 passed 0/8, 4/8 and 1/8. A take passes with Whisper coverage ≥ 0.95, at most one
extra word and a length ≤ 1.6× the expected one. After epoch 6 the model sometimes doesn't stop. That looks like data
size or overfitting, not the objective.

**Review note.** The `sft_12hz.py` hunk uses `self.model.talker…` inside `train()`, which is a plain function, so it
would raise `NameError`. The equivalent with `model.talker.text_projection(model.talker.model.text_embedding(...))`
works.
